### ISSUE TITLE
vnc for openAI gym

### ISSUE THREAD

**Issue by AwokeKnowing** [floydhub/dockerfiles#31]
openAI gym is working in the pytorch image, but to actually run the atari environments, it fails due to lack of x11.  What's the best way to get it to where we can see the environments.  I tried adding a desktop with vnc but it messed up the gpu support.

**Comment by houqp** [comment]
Could you send us the floydhub job URL so I can try to reproduce the error on my side?

**Comment by AwokeKnowing** [comment]
Hi I'm actually running this locally.  My goal is to be able to seemlessly transition from docker local to docker on floydhub or elsewhere.

I have seen some people say that the trick is to install my nvidia drivers with option --no-opengl-files 
https://davidsanwald.github.io/2016/11/13/building-tensorflow-with-gpu-support.html


So again, all I'm doing is essentially  nvidia-docker run -p5959:5900 floydhub/pytorch .  There I do see my two GPU's when running nvidia-smi.  But when running a simple OpenAI gym (pong) environment with xvfb I get the above error. 

Note, I build this entire machine from scratch precisely for working within Docker locally on OpenAi gym and pytorch

**Comment by houqp** [comment]
Could you create a minimal reproducible example on Floydhub? That can help reduce the time for us to trace down the root cause if we can reliably reproduce it on our side.

**Comment by AwokeKnowing** [comment]
@houqp yesterday I re-installed the ubuntu from scratch, updated packages, blacklisted nouveau, installed build-essentials, and installed the nvidia runfile driver (375.86), with --no-opengl-files and saying "no" to any offers to modify xconfig.  Then I installed docker and nvidia-docker and ran floydhub/pytorch, and the nvidia drivers are loaded and no errors when running against xvfb.   

So the error part of this issue can be pinned on ubuntu ppa nvidia driver (not having option to install without messing up opengl).

However, now that everything is working correctly, the title of this issue comes into play.  Since openAI gym is a graphical environment, what is the recommended way to "see" the environments rendered when running floydhub/pytorch locally (ie with a Monitor).  Does it have vnc installed? How to run it? How to connect from host OS?

Obviously I can make my own image on top of this, but is there already a vnc setup in the image?

**Comment by AwokeKnowing** [comment]
Note: here I posted a question describing the exact setup (just 2 commands on top of floydhub/pytorch)
https://askubuntu.com/questions/942768/how-can-i-properly-run-openai-gym-with-nvidia-docker-and-see-the-environments

You can see that the rendering is wrong.  Which underscores the need for floydhub to have that "already set up" in the image.

I would truly appreciate if someone with experience in installing all the packages included this.

I suggest a new image floydhub/pytorch-gui
FROM: floydhub/pytorch
[stuff to run ai Gym, matplotlib etc over vnc]

Then we could just use the image, connect over vnc  and get right to work, as floydhub's goal.  It's crazy that I've spent 12 hours just trying to get a visual of openAI gym, and it's sad that each developer would need to do the same, before starting working on deep learning code.

**Comment by houqp** [comment]
I think it would be useful to add vnc streaming support to base image so all envrionmens will have access to it. I noticed you also installed fluxbox, is that a hard dependency? 

Sorry that we are working on releasing a major update to the platform, so I won't have enough time to work on this in the upcoming weeks. That said, you are welcome to submit a patch to the dockerfile at https://github.com/floydhub/dockerfiles/blob/master/dl/dl-base/dl-base-1.x.x.jinja. I can help test and build the new image for you.

**Comment by AwokeKnowing** [comment]
@houqp 
fluxbox is not a hard dependency. I just found some example that used it. I have posted on several forums including askubuntu and devtalk.nvidia.com regarding it, but have not yet found the best way.  I supposed a very lightweight desktop in the base might be the best way.

**Comment by houqp** [comment]
I have a feeling that there is still space for optimization, we can probably get away without a desktop environment.

**Comment by AwokeKnowing** [comment]
@houqp So I got sidetracked from this project doing the Coursera DeepLearning Specialization, but when that's done I really want to nail the stack for local, docker-based OpenAI gym development (especially atari).

You're right that a full desktop is overkill. We just need working VNC of the gym environment.

**Comment by houqp** [comment]
Sounds good, looking forward to work with you on that when you finish the class. This will benefit a lot of people :)

**Comment by AwokeKnowing** [comment]
So based on the info at openAI for universe, I did the following:

```
 nvidia-docker run --privileged --rm -it \
 -e DOCKER_NET_HOST=172.17.0.1 \
 -v /var/run/docker.sock:/var/run/docker.sock \
 floydhub/pytorch:0.1.11-gpu-py3.6 bash
```
it seems to be important to map the docker socket in there for universe to start it's own docker containers.

With this setup I can run the environments and they appear in vnc localhost:5900 or web localhost:15900 (they have a built in web-based vnc client).

as an example, when running the above container, then I can write the below example code in a python script and run it, and view it from my machine at that web address.

```
import gym
import universe  # register the universe environments

env = gym.make('gym-core.Pong-v3')
env.configure(remotes=1)  # automatically creates a local docker container
observation_n = env.reset()

while True:
  action_n = [[('KeyEvent', 'ArrowUp', True)] for ob in observation_n]  # your agent here
  observation_n, reward_n, done_n, info = env.step(action_n)
  env.render()
```
![image](https://user-images.githubusercontent.com/2770322/30528016-a9769424-9be3-11e7-9005-9e51e6d3a3c7.png)

However, this is all for the **'universe'** version of the atari environments, not the native gym one.   The difference is that the universe ones are wrapped in docker and the observation comes over vnc, so the frame rate is capped and there is network overhead and so training is slower, and the reward signal has lag. So I still would like to be able to do the same with the native gym atari environments, which are also already installed the pytorch image.

But this is a good start. Pytorch + atari in one floydhub image.

**Comment by AwokeKnowing** [comment]
Ok, so I now believe the best way for local development will be x forwarding.  No need to install anything else in the image.

However, I believe something in the floydhub stack has corrupted the normal rendering.  Other environments work but all the atari ones render with corrupted display.  

To reproduce, run:
```
nvidia-docker run --privileged --rm -it --name ataritest \
    -e DOCKER_NET_HOST=172.17.0.1 \
    -v /var/run/docker.sock:/var/run/docker.sock \
    --user=$(id -u) \
    --env="DISPLAY" \
    --workdir="/home/$USER" \
    --volume="/etc/group:/etc/group:ro" \
    --volume="/etc/passwd:/etc/passwd:ro" \
    --volume="/etc/shadow:/etc/shadow:ro" \
    --volume="/etc/sudoers.d:/etc/sudoers.d:ro" \
    --volume="/tmp/.X11-unix:/tmp/.X11-unix:rw" \
    floydhub/pytorch:0.1.11-gpu-py3.6 bash
```
now in the image, type `python` and the following:
```
import gym
env = gym.make('Pong-v0')
env.reset()
env.render()
```
That should open up an x-forwarded window on your machine, but the display is corrupt (at least for me)
![image](https://user-images.githubusercontent.com/2770322/30529322-2dc47838-9bf1-11e7-84e6-0882662ab19f.png)

Above I actually used SpaceInvaders-v0

So, the point is that there is no need to install a 'desktop' or vnc or anything.  x-forwarding can allow local development viewing of the windows just fine.  However, something with openGL or something is messed up in the floydhub/pytorch image.

Please try the above and let me know if you can reproduce it.  If it works fine for you then I guess something is wrong with my drivers.

**Comment by houqp** [comment]
This is awesome, I am going to look into it today.

**Comment by houqp** [comment]
I am able to reproduce the garbled screen on my laptop. One thing I noticed though is `CartPole-v0` is working fine. So this is not happening to all the games.

**Comment by houqp** [comment]
On the other hand, i am able to save content of screen as image locally with correct content:

```python
env = gym.make('Pong-v0') 
env.reset()
env.env.ale.saveScreenPNG('test_image.png')
```

**Comment by AwokeKnowing** [comment]
in the ALE manual, I found this:
![image](https://user-images.githubusercontent.com/2770322/30794059-e2312e90-a17a-11e7-84e8-4af375d84327.png)

So ALE's default is not RGB but palette. I'm thinking maybe somehow the pallets values are getting sent to the 'observation' that gym gets, instead of rgb pixels. and gym is passing the palette values to the VNC render as if they were RGB values.  That would explain why each line seems to repeat about 3 times and the colors are wrong. 

I wonder if gym or ALE are using different 'default' settings when they run in docker vs normal install.

**Comment by houqp** [comment]
Yeah, I think the next step is to skip the gym abstraction and troubleshoot this with vanilla ALE.

**Comment by sagelywizard** [comment]
Any update? Running into the same issue.

**Comment by houqp** [comment]
Unfortunately, I haven't had time to dig into it. @sagelywizard are you interested in helping?

**Comment by AwokeKnowing** [comment]
@houqp @sagelywizard I haven't had a chance to get back to this as I've been doing the 3rd and 4th courses in the Andrew Ng Deep Learning Specialization in coursera.

However, I did actually try installing gym on the official pytorch image, and that worked with no graphics glitches.  So I don't think the solution is in troubleshooting ALE directly, but more like divide and conquer removing other dependencies in the image.

see this thread:
https://github.com/ikostrikov/pytorch-a2c-ppo-acktr/issues/9

But it confirms we shouldn't need any 'desktop' or anything special to get it working.  It will be really nice to have that working on these floydhub images because they are so well organized and have the other needed tools already built in.

**Comment by houqp** [comment]
@AwokeKnowing this is very useful info, thanks! We can start with https://hub.docker.com/r/floydhub/dl-base image and see if the glitches still happens at the base layer.

**Comment by AwokeKnowing** [comment]
@houqp So I finished the DeepLearning specialization and looked back at this issue.  Now I have found the sad reality.   It turns out that the openai universe team seems to not be working so closely with the open ai gym team.  

It turns out that universe requires a very old version of gym.  As in there have been a number of breaking changes since then.   So if you update gym, then the graphics issue is completely fixed and everything works fine with x-forwarding the atari gym env.  However, updating gym breaks universe. universe is not compatible with the updated gym.

I'm not sure how you would like to handle that.   It turns out there's more work on gym than on 'universe' mostly I guess because 'universe' is a much harder problem due to lag and it's more resource intensive.

So one solution would be to install the new gym and leave universe out.   then you could have a separate universe image (perhaps on top of the pytorch image which has about everything).  

or possibly you could figure out to do it in a virtualenv and perhaps have an -e GYM=universe environment flag which somehow selects whether you want gym to be the current gym or the 'very' old gym for universe. 

but overall, I believe that at this point having up-to-date gym is more inline with current trends. perhaps (or perhaps not) they will eventually refactor universe, but my expectation from what I have read is that it would probably be a significant refactor and is not coming soon.

**Comment by houqp** [comment]
Thanks @AwokeKnowing for digging into the root cause! I agree with you. If universe is not active anymore, we should exclude from the default package list. We are about to rebuild all images for a newer version of cuda. Will incorporate your change with the new release.

**Comment by houqp** [comment]
@AwokeKnowing we released a new set of images (pythorch-0.3.1 and tensorflow-1.7) last week with universe removed. Do you mind give it a try? I will test this on a linux machine with GUI environment later today.
